metrics-operator/api/v1alpha3/analysis_types.go

```diff
@@ -81,9 +85,58 @@ type AnalysisList struct {
 
 type Timeframe struct {
 	// From is the time of start for the query, this field follows RFC3339 time format
-	From metav1.Time `json:"from"`
+	From metav1.Time `json:"from,omitempty"`
 	// To is the time of end for the query, this field follows RFC3339 time format
-	To metav1.Time `json:"to"`
+	To metav1.Time `json:"to,omitempty"`
+	// Recent describes a recent timeframe using a duration string. E.g. Setting this to '5m' will cause the Analysis
+	// to be performed for the last five minutes
+	// +optional
+	// +kubebuilder:validation:Pattern="^0|([0-9]+(\\.[0-9]+)?(ns|us|µs|ms|s|m|h))+$"
+	// +kubebuilder:validation:Type:=string
+	// +optional
+	Recent metav1.Duration `json:"recent,omitempty"`
+}
+
+// GetFrom returns the 'from' timestamp from the status of the Analysis.
+// This function has been added to provide a clear way of retrieving the correct timestamp
+// to use, which is the one from the Status.
+func (a *Analysis) GetFrom() time.Time {
+	return a.Status.Timeframe.GetFrom()
+}
+
+// GetTo returns the 'from' timestamp from the status of the Analysis.
+// This function has been added to provide a clear way of retrieving the correct timestamp
+// to use, which is the one from the Status.
+func (a *Analysis) GetTo() time.Time {
+	return a.Status.Timeframe.GetTo()
+}
+
+func (a *Analysis) EnsureTimeframeIsSet() {
+	// make sure the correct time frame is set in the status - once an Analysis with a duration string specifying the
+	// time frame is triggered, the time frame derived from that duration should stay the same and not shift over the course
+	// of multiple reconciliation loops
+	if a.Status.Timeframe.From.IsZero() || a.Status.Timeframe.To.IsZero() {
+		a.Status.Timeframe.From = metav1.Time{
+			Time: a.Spec.GetFrom(),
+		}
+		a.Status.Timeframe.To = metav1.Time{
+			Time: a.Spec.GetTo(),
+		}
+	}
+}
+
+func (t *Timeframe) GetFrom() time.Time {
+	if t.Recent.Duration > 0 {
```

Comment:
This allows both combinations to be present. I would add a doc entry in the struct definition about the precedence of this field against from and to

Reply:
After reading the validation webhook, I see that you had extra protection in the code :) 
I will then document the fact that either from/to is used or recent

Reply:
sure, will highlight that in the documentation